packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -475,41 +500,66 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
-        // If we are not loaded, we will update this when we are realized.
         if (this.loaded) {
-            this.updateChannelUsedRoutes();
+            this.updateChannelUsedRoutes(usedRoutes, gcTimestamp);
+        } else {
+            this.pendingUsedState = { usedRoutes, gcTimestamp };
+        }
+    }
+
+    /**
+     * Updates pending GC state of the channel, if any. There can be pending state when GC runs for this data store
+     * before it is loaded.
+     */
+    private updateChannelPendingGCState() {
+        if (this.pendingUsedState !== undefined) {
+            this.updateChannelUsedRoutes(this.pendingUsedState.usedRoutes, this.pendingUsedState.gcTimestamp);
+        }
+        if (this.pendingRoutesToReset !== undefined) {
+            this.resetChannelUnreferencedState(this.pendingRoutesToReset);
         }
     }
 
     /**
-     * Updates the used routes of the channel and its child contexts. The channel must be loaded before calling this.
+     * Updates the used routes of the channel's child contexts. The channel must be loaded before calling this.
      * It is called in these two scenarions:
      * 1. When the used routes of the data store is updated and the data store is loaded.
-     * 2. When the data store is realized. This updates the channel's used routes as per last GC run.
+     * 2. When the data store is realized and there was a GC run before that.
      */
-    private updateChannelUsedRoutes() {
+    private updateChannelUsedRoutes(usedRoutes: string[], gcTimestamp?: number) {
         assert(this.loaded, 0x144 /* "Channel should be loaded when updating used routes" */);
         assert(this.channel !== undefined, 0x145 /* "Channel should be present when data store is loaded" */);
 
-        // If there is no lastUsedState, GC has not run up until this point.
-        if (this.lastUsedState === undefined) {
-            return;
-        }
+        // Remove the route to this data store, if it exists.
+        const usedChannelRoutes = usedRoutes.filter(
+            (id: string) => { return id !== "/" && id !== ""; },
+        );
+        this.channel.updateUsedRoutes(usedChannelRoutes, gcTimestamp);
+    }
+
+    /**
+     * Reset unreferenced state of the channel's child contexts. The channel must be loaded before calling this.
+     * It is called in these two scenarions:
+     * 1. When the unreferenced state of the data store is reset and the data store is loaded.
+     * 2. When the data store is realized and there was a GC run before that.
+     */
+    private resetChannelUnreferencedState(routes: string[]) {
+        assert(this.loaded, "Channel should be loaded when resetting unreferenced state");
+        assert(this.channel !== undefined, "Channel should be present when data store is loaded");
 
         // Remove the route to this data store, if it exists.
-        const usedChannelRoutes = this.lastUsedState.usedRoutes.filter(
+        const channelResetRoutes = routes.filter(
```

Comment:
Can this filtering happen in caller (resetUnreferencedState) such that there is no need to have that function - both callers could call this.channel.resetUnreferencedState directly